add v2 support

liubin opened the pull request.

liubin committed:
probe cgroup v1 mount path

Signed-off-by: bin liu <bin@hyper.sh>

liubin committed:
copy cpuset after created

Signed-off-by: bin liu <bin@hyper.sh>

liubin committed:
add hugetlb functions

Signed-off-by: bin liu <bin@hyper.sh>

liubin committed:
add disable oom_control function

Signed-off-by: bin liu <bin@hyper.sh>

liubin committed:
change *limit* in memory from u64 to i64

Signed-off-by: bin liu <bin@hyper.sh>

liubin committed:
fix: create all dirs

Signed-off-by: bin liu <bin@hyper.sh>

liubin committed:
add basic v2 cpu/memory functions

Signed-off-by: bin liu <bin@hyper.sh>

liubin committed:
add systemd controller; add more tests; update blkio stat for v2

Signed-off-by: bin liu <bin@hyper.sh>

liubin committed:
get relative paths at new/load func default.

Signed-off-by: bin liu <bin@hyper.sh>

liubin committed:
add SPDX-License-Identifier for all source

Add SPDX-License-Identifier for all existing codes.

Fixes: #2

Signed-off-by: bin liu <bin@hyper.sh>

bergwolf committed:
Merge pull request #3 from liubin/fix/2

add SPDX-License-Identifier for all source

liubin commented on src/blkio.rs:
```diff
@@ -662,16 +770,27 @@ impl BlkIoController {
+        let mut content = format!("{}:{} {}", major, minor, iops);
+        if self.v2 {
+            file = "io.max";
+            content = format!("{}:{} riops={}", major, minor, iops);
```
这里应该是 wriops

liubin commented on src/blkio.rs:
```diff
@@ -685,7 +804,13 @@ impl BlkIoController {
+        let mut file = "blkio.weight_device";
+        if self.v2 {
+            // FIXME why is there no weight for device in runc ?
+            // https://github.com/opencontainers/runc/blob/46be7b612e2533c494e6a251111de46d8e286ed5/libcontainer/cgroups/fs2/io.go#L30
```
好像跟 io 调度算法有关 。https://wiki.ubuntu.com/Kernel/Reference/IOSchedulers

liubin commented on tests/devices.rs:
```diff
@@ -1,12 +1,18 @@
 fn test_devices_parsing() {
-    let hier = cgroups::hierarchies::V1::new();
-    let cg = Cgroup::new(&hier, String::from("test_devices_parsing"));
+    // no only v2
```
no -> now

liubin commented on tests/hugetlb.rs:
```diff
@@ -0,0 +1,39 @@
+
+#[test]
+fn test_hugetlb_sizes() {
+    // no only v2
```
no -> now

lifupan commented on src/blkio.rs:
```diff
@@ -269,17 +329,29 @@ impl ControllerInternal for BlkIoController {
         if res.update_values {
-            let _ = self.set_weight(res.weight as u64);
-            let _ = self.set_leaf_weight(res.leaf_weight as u64);
+            if res.weight.is_some() {
```
这个地方的几个 判断 is_some 的句式 改成下面的样子吧：
if let Some(weight) = res.weight {
     let _ = self.set_weight(weight as u64);
}

liubin committed:
use let Some() instead of is_some() to make codes clear

Signed-off-by: bin liu <bin@hyper.sh>

liubin committed:
Merge remote-tracking branch 'upstream/master'

liubin committed:
add SPDX-License-Identifier for modified files

Signed-off-by: bin liu <bin@hyper.sh>

liubin committed:
cargo fmt

Signed-off-by: bin liu <bin@hyper.sh>

liubin committed:
hugetlb: update test to avoid depending on test env

Signed-off-by: bin liu <bin@hyper.sh>

bergwolf committed:
Merge pull request #4 from liubin/master

add cgroup v2 support

liubin committed:
cpu: change cfs_quota from u64 to i64

cfs_quota can be set to -1 to indicate no limit

Fixes: #5

Signed-off-by: bin liu <bin@hyper.sh>

liubin committed:
errors: add sync marker to error cause

Errors cause should be shared to let anyhow to wrap it.

Fixes: #7

Signed-off-by: bin liu <bin@hyper.sh>

bergwolf committed:
Merge pull request #6 from liubin/fix/cfs_quota-to-i64

cpu: change cfs_quota from u64 to i64

bergwolf committed:
Merge pull request #8 from liubin/fix/7-add-sync-marker-to-error-cause

errors: add sync marker to error cause

Kvasscn committed:
lib: fix compiling error for type i8 in aarch64 environment

c_char is equal to i8 on x86, but it is equal to u8 on arm arch,
so we need to use type c_char instead of i8 to avoid compiling error.

Fixes: #9

Signed-off-by: zhanghj <zhanghj.lc@inspur.com>

lifupan committed:
Merge pull request #10 from Kvasscn/fix-type-i8-compile-error-in-aarch64

lib: fix compiling error for type i8 in aarch64 environment

liubin committed:
update Travis configs

Fixes: #11

Signed-off-by: bin liu <bin@hyper.sh>

bergwolf committed:
Merge pull request #12 from liubin/fix/11-add-trivas

update Travis configs
